bmm/src/main/java/org/openehr/bmm/core/BmmModel.java

```diff
@@ -154,13 +147,87 @@ public boolean isMsConformantPropertyType(String bmmTypeName, String bmmProperty
+        if (bmmClass != null) {
+            return hasPropertyAtPath(bmmClass, new APathQuery(propertyPath));
+        } else {
+            return false;
+        }
+    }
+
+    private boolean hasPropertyAtPath (BmmClass bmmClass, APathQuery qPath) {
+        int pathPos = qPath.index();
+        boolean result = false;
+        if (bmmClass.hasFlatPropertyWithName(qPath.itemName())) {
+            if (qPath.isLast()) {
+                result = true;
+            } else {
+                BmmClass bmmPropTypeClass = getClassDefinition(bmmClass.getFlatProperties().get(qPath.itemName()).getType().getEffectiveType().typeBaseName());
+                qPath.forth();
+                if (bmmPropTypeClass != null) {
+                    result = hasPropertyAtPath(bmmPropTypeClass, qPath);
+                }
+            }
+        } else {
+            for (String descClass : bmmClass.getImmediateDescendants()) {
+                if (hasPropertyAtPath(getClassDefinition(descClass), qPath)) {
+                    result = true;
+                    break;
+                }
+            }
+        }
+        qPath.go (pathPos);
+        return result;
+    }
+
+    /**
+     * Return BmmProperty object at a path within a type, including among flat properties
+     * and descendant classes (i.e. compute dynamic attachment possibilities).
+     * @precondition hasPropertyAtPath ()
+     * @param typeName : type within which to find the property
+     * @param propertyPath : path
+     * @return BmmProperty
+     */
+    public BmmProperty propertyAtPath (String typeName, String propertyPath) {
+        BmmClass bmmClass = getClassDefinition(typeName);
+        if (bmmClass != null) {
+            return propertyAtPath (bmmClass, new APathQuery(propertyPath));
+        } else {
+            return null;
+        }
+    }
+
+    private BmmProperty propertyAtPath (BmmClass bmmClass, APathQuery qPath) {
+        BmmProperty result = null;
+        int pathPos = qPath.index();
+        if (bmmClass.hasFlatPropertyWithName (qPath.itemName())) {
+            if (qPath.isLast()) {
+                result = bmmClass.getFlatProperties().get(qPath.itemName());
+            } else {
+                BmmClass bmmPropTypeClass = getClassDefinition (bmmClass.getFlatProperties().get(qPath.itemName()).getType().getEffectiveType().typeBaseName());
+                qPath.forth();
+                if (bmmPropTypeClass != null) {
+                    result = propertyAtPath (bmmPropTypeClass, qPath);
+                }
+            }
+        } else {
+            for (String descClass : bmmClass.getImmediateDescendants()) {
+                BmmProperty descClassProperty = propertyAtPath (getClassDefinition(descClass), qPath);
+                if (descClassProperty != null) {
+                    result = descClassProperty;
+                    break;
+                }
+            }
+        }
+        qPath.go (pathPos);
+        return result;
     }
```

Comment:
This can also be done without the custom iterator in `APathQuery`:
```suggestion
    public BmmProperty propertyAtPath (String typeName, String propertyPath) {
        BmmClass bmmClass = getClassDefinition(typeName);
        if (bmmClass != null) {
            return propertyAtPath (bmmClass, new APathQuery(propertyPath).getPathSegments());
        } else {
            return null;
        }
    }

    private BmmProperty propertyAtPath (BmmClass bmmClass, List<PathSegment> pathSegments) {
        BmmProperty result = null;
        String nodeName = pathSegments.get(0).getNodeName();
        if (bmmClass.hasFlatPropertyWithName (nodeName)) {
            if (pathSegments.size() == 1) {
                result = bmmClass.getFlatProperties().get(nodeName);
            } else {
                BmmClass bmmPropTypeClass = getClassDefinition (bmmClass.getFlatProperties().get(nodeName).getType().getEffectiveType().typeBaseName());
                if (bmmPropTypeClass != null) {
                    result = propertyAtPath (bmmPropTypeClass, pathSegments.subList(1, pathSegments.size()));
                }
            }
        } else {
            for (String descClass : bmmClass.getImmediateDescendants()) {
                BmmProperty descClassProperty = propertyAtPath (getClassDefinition(descClass), pathSegments);
                if (descClassProperty != null) {
                    result = descClassProperty;
                    break;
                }
            }
        }
        return result;
    }
```

Reply:
I agree. Sublist does not copy the data, so still fast. Alternative is using the built-in qpath.getPathSegments().iterator() instead of creating a custom one.